### PR TITLE
Add support for async job handling

### PR DESCRIPTION
**Pull request by suneeth51**
- The SDK handles the the async job id and retries until job_timeout expires with interval of 5 seconds. If job doesn't complete within timeout, then client timeout exception with the job status is returned back to the user.
- By default, the SDK client supports default timeout of 60 seconds. However, user can override to specific timeout value both at the client level and per request level.
- The request timeout will takes precedence if specified.

### CODE REVIEW

File: nimbleclient/resource.py [hunk 1]
```diff
@@ -62,15 +62,15 @@ def get(self, id=None, **kwargs):
                 return self.resource(objs[0]['id'] if 'id' in objs[0] else None, objs[0], client=self._client, collection=self)
 
     def create(self, name, **kwargs):
-        resp = self._client.create_resource(self.resource_type, name=name, **kwargs)
+        resp = self._client.create_resource(self.resource_type, name, **kwargs)
```
Comment (cristian-medina) [inline nimbleclient/resource.py]: keep `name=name` here

File: nimbleclient/v1/restclient.py [hunk 2]
```diff
@@ -392,3 +433,38 @@ def perform_bulk_resource_action(self, resource, action, **params):
...
+
+        now = datetime.datetime.now()
+        expiry_time = now + datetime.timedelta(0, timeout_sec)
+        logging.debug(f"job_id: [{job_id}], job_timeout: [{timeout_sec} seconds], current_time: [{now}], job_retry_expiry_time: [{expiry_time}], retry_interval: [{retry_interval} seconds]")
+        while 1:
+            try:
+                # Get job status from Array
+                response = self.get_resource('jobs', job_id)
+                logging.debug("JOB STATUS RESPONSE: %s", json.dumps(response, indent=4))
+                if (response['state'] == '1done'):
```
Comment (cristian-medina) [inline nimbleclient/v1/restclient.py]: Double check that this conditional is not a typo. Is it `1done` not `done`?

**Review by cristian-medina — CHANGES_REQUESTED** [review]

File: nimbleclient/v1/restclient.py (hunk 2 above)
Comment (suneeth51) [reply nimbleclient/v1/restclient.py]: oh right.. I will change that to 'done'. Thanks!

File: nimbleclient/resource.py (hunk 1 above)
Comment (suneeth51) [reply nimbleclient/resource.py]: Agreed, I will update it.

**Review by cristian-medina — APPROVED** [review]

File: nimbleclient/v1/restclient.py [hunk 3]
```diff
@@ -268,6 +281,14 @@ def delete(self, endpoint):
                         self._refresh_connection()
                     else:
                         raise NimOSAPIError(response.json())
+                elif response.status_code == 202:
+                    # Handle async job response
+                    job_response = response.json()
+                    if 'messages' in job_response and job_response['messages'][0]['code'] == 'SM_async_job_id':
+                        logging.debug("Got async job response: %s", json.dumps(job_response['messages'], indent=4))
```
Comment (ar-india) [inline nimbleclient/v1/restclient.py:288]: hope we are not dumping any sensitive data here.. Please take care of this.. if yes then we may want to remove the logging

File: nimbleclient/v1/restclient.py [hunk 4]
```diff
@@ -299,6 +322,14 @@ def put(self, endpoint, **payload):
...
```
Comment (ar-india) [inline nimbleclient/v1/restclient.py:329]: same here

File: nimbleclient/v1/restclient.py [hunk 5]
```diff
@@ -392,3 +433,38 @@ def perform_bulk_resource_action(self, resource, action, **params):
...
+                if datetime.datetime.now().time() > expiry_time.time():
+                    logging.error("Job status retry timeout has expired after [%s] seconds, returning with client timeout exception", timeout_sec)
+                    raise NimOSClientJobTimeoutError(response)
+                else:
+                    #  Sleep n Wait
+                    if timeout_sec < retry_interval:
+                        logging.debug("Sleeping for [%s seconds] before next retry attempt ...", timeout_sec)
+                        time.sleep(timeout_sec)
+                    else:
+                        logging.debug("Sleeping for [%s seconds] before next retry attempt ...", retry_interval)
```
Comment (ar-india) [inline nimbleclient/v1/restclient.py:463]: where are we reducing the retry_interval?? it is always set to 5

Comment (ar-india) [reply nimbleclient/v1/restclient.py:463]: we will be running in infinite loop

**Pull request merged**